File: apps/windows/src/Polyglance.Core/Services/TextFormattingService.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Text.Json;
using Polyglance.Core.Models;
using Polyglance.Core.Native;

namespace Polyglance.Core.Services;

/// <summary>
/// How OCR output is reflowed before it reaches the clipboard or a translation
/// request. The numeric values are persisted in the user's configuration and
/// must stay in step with <c>capture_core::formatting::TextFormattingMode</c>.
/// </summary>
public enum TextFormattingMode
{
    SmartMerge = 0,
    PreserveBreaks = 1,
    RemoveSpaces = 2,
    Raw = 3
}

/// <summary>
/// Thin wrapper over the shared Rust implementation in
/// <c>capture_core::formatting</c>. The algorithm itself lives there so macOS
/// and Windows cannot drift apart; only the P/Invoke marshalling is local.
/// </summary>
public static class TextFormattingService
{
    public static string Format(IEnumerable<LayoutTextLine> lines, TextFormattingMode mode)
    {
        if (lines == null)
            return string.Empty;

        var list = lines as IList<LayoutTextLine> ?? lines.ToList();
        if (list.Count == 0)
            return string.Empty;

        string linesJson = JsonSerializer.Serialize(list);
        int status = NativeMethods.polyglance_layout_format_text(linesJson, (byte)mode, out IntPtr outText);
        if (status != 0 || outText == IntPtr.Zero)
        {
            string plainText = string.Join("\n", list.Select(l => l.Text));
            return Format(plainText, mode);
        }

        try
        {
            return Marshal.PtrToStringUTF8(outText) ?? string.Empty;
        }
        finally
        {
            NativeMethods.polyglance_free_string(outText);
        }
    }

    public static string Format(string text, TextFormattingMode mode)
    {
        if (string.IsNullOrEmpty(text))
            return string.Empty;

        return Transform(
            text,
            (string input, out IntPtr output) =>
                NativeMethods.polyglance_text_format(input, (byte)mode, out output));
    }

    public static string SmartMergeLines(string text)
    {
        if (string.IsNullOrEmpty(text))
            return string.Empty;

        return Transform(text, NativeMethods.polyglance_text_smart_merge_lines);
    }

    public static string ApplyPanguSpacing(string text)
    {
        if (string.IsNullOrEmpty(text))
            return string.Empty;

        return Transform(text, NativeMethods.polyglance_text_apply_pangu_spacing);
    }

    public static string RemoveExtraneousSpaces(string text)
    {
        if (string.IsNullOrEmpty(text))
            return string.Empty;

        return Transform(text, NativeMethods.polyglance_text_remove_extraneous_spaces);
    }

    /// <summary>
    /// A UTF-16 code unit outside the Basic Multilingual Plane is half of a
    /// surrogate pair and never CJK on its own, which matches how the callers
    /// inspect one <c>char</c> at a time.
    /// </summary>
    public static bool IsCjk(char value) =>
        NativeMethods.polyglance_text_is_cjk_scalar(value);

    private delegate int NativeTransform(string text, out IntPtr outText);

    /// <summary>
    /// Returns the input unchanged if the native call fails, so a formatting
    /// problem degrades to raw OCR text instead of losing it.
    /// </summary>
    private static string Transform(string text, NativeTransform operation)
    {
        int status = operation(text, out IntPtr outText);
        if (status != 0 || outText == IntPtr.Zero)
        {
            return text;
        }

        try
        {
            return Marshal.PtrToStringUTF8(outText) ?? text;
        }
        finally
        {
            NativeMethods.polyglance_free_string(outText);
        }
    }
}

File: apps/macos/Sources/PolyglanceKit/TextFormattingService.swift
import Foundation
import TranslatorCore

public enum TextFormattingMode: Int, CaseIterable, Sendable, Codable {
    case smartMerge = 0
    case preserveBreaks = 1
    case removeSpaces = 2
    case raw = 3

    public var title: String {
        switch self {
        case .smartMerge:
            return "智能合并折行"
        case .preserveBreaks:
            return "保持原样换行"
        case .removeSpaces:
            return "清除多余空格"
        case .raw:
            return "原始文本"
        }
    }
}

/// Thin wrapper over `capture_core::formatting`, so macOS and Windows reflow OCR
/// text identically. Every method forwards to the shared Rust core.
public enum TextFormattingService {
    public static func format(_ text: String, mode: TextFormattingMode) -> String {
        textFormat(text: text, mode: UInt8(mode.rawValue))
    }

    public static func format(lines: [(text: String, boundingBox: CGRect)], mode: TextFormattingMode) -> String {
        guard !lines.isEmpty else { return "" }
        let uniffiLines = lines.map { item in
            LayoutTextLine(
                text: item.text,
                boundingBox: CaptureRect(
                    x: Double(item.boundingBox.origin.x),
                    y: Double(item.boundingBox.origin.y),
                    width: Double(item.boundingBox.size.width),
                    height: Double(item.boundingBox.size.height)
                )
            )
        }
        return layoutFormatText(lines: uniffiLines, mode: UInt8(mode.rawValue))
    }

    public static func smartMergeLines(_ text: String) -> String {
        textSmartMergeLines(text: text)
    }

    public static func applyPanguSpacing(_ text: String) -> String {
        textApplyPanguSpacing(text: text)
    }

    public static func removeExtraneousSpaces(_ text: String) -> String {
        textRemoveExtraneousSpaces(text: text)
    }

    /// Swift compares `Character` values, so the first scalar of the grapheme
    /// cluster decides CJK-ness, matching what the callers already assumed.
    public static func isCjk(_ character: Character) -> Bool {
        guard let scalar = character.unicodeScalars.first else { return false }
        return textIsCjkScalar(scalar: scalar.value)
    }
}
